### PR TITLE
[FIX] properly paste/drop escaped text and filtered html

### PR DESCRIPTION
**Pull request by Zinston**
Thus far, the editor only handled text-only pasting/dropping. This commit makes it so that it now handles html pasting/dropping, and ensures that text pasting is properly escaped (to prevent evaluating html strings on paste).
When pasting html, the editor filters the nodes, their attributes and their classes through a combination of a whitelist and a blacklist. Nodes that match blacklisted or non-whitelisted selectors are removed and their contents are unwrapped recursively. Attributes and classes that are blacklisted or non-whitelisted are removed. What is kept are basic styling elements (eg: h1, strong, ...), images (with their src attribute), links (with the href attribute), and a selection of classes the edition of which is supported by the editor.

### CODE REVIEW

File: src/editor.js [hunk 1]
```diff
@@ -1257,6 +1296,124 @@ export class OdooEditor extends EventTarget {
...
+     * `justContainer` is true) for safely pasting. Cleaning an element involves
+     * unwrapping its contents if it's an illegal (blacklisted or not
+     * whitelisted) element, or removing its illegal attributes and classes.
+     * Return the cleaned element.
+     *
+     * @param {Node} container
+     * @param {boolean} [justContainer=false]
+     * @returns {Node}
+     */
+    _cleanForPaste(container, justContainer = false) {
```
Comment (Zinston) [inline src/editor.js:1352]: Not a fan of `justContainer` but can't think of anything better.

I could maybe pass an array of nodes to `_cleanForPaste` but I'm afraid that would look more complicated. Not sure though.

File: src/editor.js [hunk 2]
```diff
@@ -1500,16 +1657,20 @@ export class OdooEditor extends EventTarget {
...
+            this.execCommand('insertHTML', this._prepareClipboardData(clipboardData));
+        } else {
+            const text = ev.clipboardData.getData('text/plain');
+            this.execCommand('insertHTML', this._prepareTextClipboardData(text));
+        }
     }
-
     /**
-     * Prevent the dropping of HTML and paste text only instead.
+     * Handle safe dropping of html into the editor.
```
Comment (Zinston) [inline src/editor.js:1673]: I noticed that you can't seem to actually be able to drop plain text. Whatever I dropped was always html. Text just won't drop.

File: src/editor.js [hunk 3]
```diff
@@ -1500,16 +1657,20 @@ export class OdooEditor extends EventTarget {
...
-     * Prevent the pasting of HTML and paste text only instead.
+     * Handle safe pasting of html or plain text into the editor.
      */
     _onPaste(ev) {
         ev.preventDefault();
-        const pastedText = (ev.originalEvent || ev).clipboardData.getData('text/plain');
-        this.execCommand('insertText', pastedText);
+        const clipboardData = ev.clipboardData.getData('text/html');
+        if (clipboardData) {
+            this.execCommand('insertHTML', this._prepareClipboardData(clipboardData));
```
Comment (Zinston) [inline src/editor.js:1666]: If I drop a <p> inside a <p> it will indeed end up with <p><p>, which it shouldn't. This is a bug of `insertHTML` that should be addressed in another PR.

File: src/editor.js [hunk 4]
```diff
@@ -1257,6 +1296,124 @@ export class OdooEditor extends EventTarget {
...
+    /**
+     * Prepare clipboard data (text/html) for safe pasting into the editor.
+     *
+     * @private
+     * @param {string} clipboardData
+     * @returns {string}
+     */
+     _prepareClipboardData(clipboardData) {
+        const container = document.createElement('fake-container');
+        container.innerHTML = clipboardData;
```
Comment (dmo-odoo) [inline src/editor.js:1310]: Security is going to scream. Paste html with a script in it and poof, it's executed now.

File: src/editor.js (hunk 1 above)
Comment (dmo-odoo) [reply src/editor.js:1352]: I don't understand why this `justContainer` parameter is necessary. Is the function recursive if and only if this parameter is `true` ? If yes, then call it `recurse` or something like that. If not, then why not simply loop over the elements children at call site when you would call the function with `false` and call it for each child instead ?

File: src/editor.js [hunk 5]
```diff
@@ -1257,6 +1296,124 @@ export class OdooEditor extends EventTarget {
...
+     * Return the cleaned element.
+     *
+     * @param {Node} container
+     * @param {boolean} [justContainer=false]
+     * @returns {Node}
+     */
+    _cleanForPaste(container, justContainer = false) {
+        for (const node of justContainer ? [container] : [...container.childNodes]) {
+            if (!this._isWhitelisted(node) || this._isBlacklisted(node)) {
+                if (['META', 'STYLE', 'SCRIPT'].includes(node.nodeName)) {
```
[the remark this answers is not in the document]
Comment (dmo-odoo) [reply src/editor.js:1355]: I don't understand what you mean. However, I'm in favor of having blacklist as an object if only for the sake of consistency.

File: src/editor.js [hunk 6]
```diff
@@ -1257,6 +1296,124 @@ export class OdooEditor extends EventTarget {
...
+     */
+    _isWhitelisted(item) {
+        if (item instanceof Attr) {
+            return CLIPBOARD_WHITELISTS.attributes.includes(item.name);
+        } else if (typeof item === 'string') {
+            return CLIPBOARD_WHITELISTS.classes.some(okClass => (
+                okClass instanceof RegExp ? okClass.test(item) : okClass === item
+            ));
+        } else {
+            return CLIPBOARD_WHITELISTS.nodes.some(selector => (
```
[the remark this answers is not in the document]
Comment (dmo-odoo) [reply src/editor.js:1398]: That actually might save you. Once this is merged, someone very smart will paste 10000 words in the editor and complain that it's ~~slow~~ utterly broken.

File: src/editor.js (hunk 4 above)
Comment (Zinston) [reply src/editor.js:1310]: Well sure... but it's adding that innerHTML to a container we don't directly add to the dom, and that we then clean. They had validated that method for we3.

File: src/editor.js (hunk 1 above)
Comment (Zinston) [reply src/editor.js:1352]: `justContainer === false` -> clean the container's children (not the container itself).
`justContainer === true` -> clean the container itself (not its children).

I also don't like it. I could instead systematically pass an array of nodes to clean (`justContainer ? [container] : [...container.childNodes]` basically).

File: src/editor.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (Zinston) [reply src/editor.js:1355]: Here I'm repeating some of the blacklisted selectors because these in particular I don't even want to keep their children, I want to nuke them. So I thought I could maybe also have two keys in my blacklist so I could identify these.
(Oh and that's funny cause it also reminded me of that debate :p If you think back on it though you'll recall that while I disagreed with you on the case of `master`, I didn't regarding `whitelist` and `blacklist` ^^ - though I'm not excluding that there's history I'm not aware of there).

File: src/editor.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (Zinston) [reply src/editor.js:1398]: Haha ok then :p

File: src/editor.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (dmo-odoo) [reply src/editor.js]: I was mostly trolling... :p I think your recollection of the debate is accurate though, my bad. I would advise for something else than `nuke` though. Probably `remove` ?

File: src/editor.js (hunk 1 above)
Comment (dmo-odoo) [reply src/editor.js]: I would have moved the loop outside of `_cleanForPaste` and rename `containers` to `node`. This way, the method always cleans the given node and that's it. If you need to clear more nodes, call it on each node. It seems more elegant to me. What do you think ?

File: src/editor.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (Zinston) [reply src/editor.js]: Sure I just played along on the troll, no worries :-P
`remove` sounds good :+1:
